🛠️ Implemented CustomGlobalExceptionHandler & added validation for Cr…

Volodymyr-Mykychak opened the pull request:
📖 Pull Request: Added Global Exception Handling & Data Validation 🚀
✨ Changes in this PR:
✅ Added CustomGlobalExceptionHandler for centralized error handling.
✅ Implemented validation for CreateBookRequestDto using @NotNull, @Size, and @DecimalMin.
✅ Refactored API error responses to return structured JSON with error details.
✅ Fixed Checkstyle issues (imports order, line length, and formatting).

Volodymyr-Mykychak committed:
🛠️ Implemented CustomGlobalExceptionHandler & added validation for CreateBookRequestDto ✅

sarakhmen commented on src/main/java/com/store/book/dto/CreateBookRequestDto.java:
```diff
@@ -1,16 +1,31 @@
     private BigDecimal price;
+    @Size(max = 500, message = "Опис книги не може бути довшим за 500 символів")
     private String description;
+    @NotBlank(message = "Посилання на обкладинку книги не може бути порожнім")
```
let's use English for messages

sarakhmen commented on src/main/java/com/store/book/dto/CreateBookRequestDto.java:
```diff
@@ -1,16 +1,31 @@
 
 @Data
-@JsonIgnoreProperties(ignoreUnknown = true)
+@JsonIgnoreProperties(ignoreUnknown = true) // Ігнорує невідомі поля у JSON
```
remove redundant comment

sarakhmen commented on src/main/java/com/store/book/exception/CustomGlobalExceptionHandler.java:
```diff
@@ -0,0 +1,29 @@
+            MethodArgumentNotValidException ex) {
+        Map<String, Object> errors = new HashMap<>();
+        errors.put("timestamp", LocalDateTime.now()); // Додаємо час помилки
+        errors.put("status", HttpStatus.BAD_REQUEST.value()); // Додаємо статус 400
```
remove comments

sarakhmen replied:
status field is redundant as it's a part of the response and you're passing it as a ResponseEntity param

sarakhmen commented on src/main/java/com/store/book/exception/CustomGlobalExceptionHandler.java:
```diff
@@ -0,0 +1,29 @@
+
+@RestControllerAdvice
+public class CustomGlobalExceptionHandler {
+    @ExceptionHandler(MethodArgumentNotValidException.class)
```
did you cover all the exceptions of your application?

sarakhmen reviewed (changes requested):
let's improve your solution :)

Volodymyr-Mykychak committed:
🐛 Fix: correct coverImage mapping and validation

🔧 Refactored exception handler:
- Unified validation responses (timestamp, status, errors)
- Removed code duplication with helper method

🧩 Fixed DTO-to-Entity mapping:
- Ensured 'coverImage' field is correctly mapped in all layers
- Verified field names match across CreateBookRequestDto, Book, and BookDto

✅ Validation:
- Resolved issue with @NotBlank coverImage failing due to null value

🛠 Minor cleanup to ensure MapStruct works properly

Elena-Bruyako reviewed (changes requested):
fix build first

Volodymyr-Mykychak committed:
🔧 Fix CI workflow: remove pull_request_target and unsafe actions 🚫

Volodymyr-Mykychak committed:
🔄 Rename CI to force GitHub update

Volodymyr-Mykychak committed:
🧹 Temp delete CI workflow to clear old state

Volodymyr-Mykychak committed:
🔁 Re-add clean CI workflow

Volodymyr-Mykychak committed:
🚫 Remove auto-reject-action to fix 403 error on forks

Elena-Bruyako commented on src/main/java/com/store/book/controller/BookController.java:
```diff
@@ -20,31 +21,26 @@
     @PostMapping
-    public BookDto save(@RequestBody CreateBookRequestDto bookDto) {
-
+    public BookDto save(@Valid @RequestBody CreateBookRequestDto bookDto) {
```
Didn't find `update` method in controller, add it

Elena-Bruyako replied:
fix comment

Elena-Bruyako replied:
Could you add implementation of update method? You should do this in previous task

Elena-Bruyako commented on src/main/java/com/store/book/controller/BookController.java:
```diff
@@ -20,31 +21,26 @@
 @RestController
 @RequestMapping(value = "/books")
```
```suggestion
@RequestMapping("/books")
```

Elena-Bruyako replied:
fix comment

Elena-Bruyako commented on src/main/java/com/store/book/dto/CreateBookRequestDto.java:
```diff
@@ -1,16 +1,31 @@
     private BigDecimal price;
+    @Size(max = 500, message = "Book description must not exceed 500 characters")
     private String description;
+    @NotBlank(message = "Book cover URL must not be empty")
```
```suggestion
```

Elena-Bruyako replied:
fix comment

Elena-Bruyako commented on src/main/java/com/store/book/exception/CustomGlobalExceptionHandler.java:
```diff
@@ -0,0 +1,73 @@
+        return buildErrorResponse(HttpStatus.BAD_REQUEST,
+                Map.of("error", "Invalid request format"));
+    }
+
```
```suggestion
```

Elena-Bruyako replied:
fix comment

Volodymyr-Mykychak committed:
♻️ Rename CI to trigger clean version without pull_request_target

Elena-Bruyako commented on src/main/java/com/store/book/exception/CustomGlobalExceptionHandler.java:
```diff
@@ -0,0 +1,73 @@
+        return buildErrorResponse(HttpStatus.CONFLICT,
+                Map.of("error",
+                        "Database integrity error: " + ex.getMostSpecificCause().getMessage()));
+    }
```
```suggestion
```

Elena-Bruyako commented on src/main/java/com/store/book/exception/CustomGlobalExceptionHandler.java:
```diff
@@ -0,0 +1,73 @@
+                                                                   Map<String, String> errors) {
+        Map<String, Object> body = new HashMap<>();
+        body.put("timestamp", LocalDateTime.now());
+        body.put("status", status.value());
```
```suggestion
```

Elena-Bruyako replied:
fix comment

Volodymyr-Mykychak replied:
Hi! Could you please clarify what exactly you mean by "fix comment"? I'm not fully sure which comments you are referring to or what needs to be updated. Thanks in advance 🙏

Volodymyr-Mykychak committed:
🆕 Add clean workflow to avoid pull_request_target bug

Elena-Bruyako reviewed (changes requested):
You haven't fixed any comments, why do you request review?

Volodymyr-Mykychak committed:
📝 Fix comments and add missing JavaDocs for exception handlers and DTOs

Elena-Bruyako reviewed (changes requested):
one comment

Volodymyr-Mykychak committed:
✨ add update book endpoint

Added `update` method to BookController, BookService and BookServiceImpl.
Now books can be updated via PUT /books/{id}.
Validated input with `@Valid` and updated model fields manually.

✅ Added:
- BookController#update
- BookService#update
- BookServiceImpl#update

🛠 No changes needed in DTO or mapper.

#spring #restapi #update #bookstore
